<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>Title</title>
</head>
<body>
<button id="btn">获取图片</button>

<script>
    const arr = [
        'http://edus.nodeing.com/files/system/block_picture_1516366943.jpg?version=8.2.15',
        'http://edu.nodeing.com/files/system/block_picture_1516369814.jpg?version=8.2.15',
        'http://edu.nodeing.com/files/system/block_picture_1516977533.jpg?version=8.2.15',
        'http://edu.nodeing.com/files/system/block_picture_1516973104.jpg?version=8.2.15'
    ];

    function AsyncLoadImg(url) {
        let p = new Promise((resolve, reject) => {
            let oImg = new Image();
            oImg.src = url;
            oImg.onload = () => {
                resolve(oImg)
            };
        });

        return p;
    }
    //图片超时测试
    function timeOut() {
        return new Promise((resolve, reject) => {
            setTimeout(() => {
                reject(new Error('图片超时'));
            }, 1000)
        })
    }

    /*race()方法：是将多个 Promise 实例，包装成一个新的 Promise 实例
    const p = Promise.race([p1, p2, p3]);
    上面代码中，只要p1、p2、p3之中有一个实例率先改变状态，p的状态就跟着改变。那个率先改变的 Promise 实例的返回值，就传递给p的回调函数。
    */
    let oBtn = document.getElementById('btn');
    oBtn.onclick = () => {
        //下面race()方法先回去调用AsyncLoadImg(arr[0])，如果请求成功，那么会执行then方法。但是我们将图片地址改为不正确，所以是请求不成功的。
        //而10秒后执行timeOut()。也就说在10秒内AsyncLoadImg(arr[0])得不到结果那么会执行catch方法。
        let p = Promise.race([AsyncLoadImg(arr[0]), timeOut()])
            .then((result) => {
                document.body.appendChild(result)
            }).catch((err) => {
                console.log(err);
            })
    }

</script>
</body>
</html>
